extensions/blocks/instagram-gallery/edit.js

```diff
@@ -1,61 +1,243 @@
 import { InspectorControls } from '@wordpress/block-editor';
 import {
 	Button,
-	Dashicon,
 	ExternalLink,
 	PanelBody,
+	PanelRow,
 	Placeholder,
 	RangeControl,
+	Spinner,
+	withNotices,
 } from '@wordpress/components';
+import { useEffect, useState } from '@wordpress/element';
 import { __ } from '@wordpress/i18n';
+import { addQueryArgs } from '@wordpress/url';
+
+/**
+ * Internal dependencies
+ */
+import PopupMonitor from 'lib/popup-monitor';
+import defaultAttributes from './attributes';
+import { getGalleryCssAttributes } from './utils';
+import { getValidatedAttributes } from '../../shared/get-validated-attributes';
+import './editor.scss';
+
+export function InstagramGalleryEdit( props ) {
+	const { attributes, className, noticeOperations, noticeUI, setAttributes } = props;
+	const { accessToken, align, columns, count, images, instagramUser, spacing } = attributes;
+
+	const [ isConnecting, setIsConnecting ] = useState( false );
+	const [ isLoadingGallery, setIsLoadingGallery ] = useState( false );
+
+	useEffect( () => {
+		const validatedAttributes = getValidatedAttributes( defaultAttributes, attributes );
+		if ( ! isEqual( validatedAttributes, attributes ) ) {
+			setAttributes( validatedAttributes );
+		}
+	}, [ attributes ] );
+
+	useEffect( () => {
+		if ( ! accessToken ) {
+			return;
+		}
+		noticeOperations.removeAllNotices();
+		if ( isEmpty( images ) ) {
+			setIsLoadingGallery( true );
+		}
+		apiFetch( {
+			path: addQueryArgs( '/wpcom/v2/instagram/gallery', {
+				access_token: accessToken,
+				count,
+			} ),
+		} ).then( response => {
+			setIsLoadingGallery( false );
+
+			if ( isEmpty( response.images ) ) {
+				noticeOperations.createErrorNotice(
+					__( 'No images were found in your Instagram account.', 'jetpack' )
+				);
+				return;
+			}
+
+			setAttributes( { images: response.images, instagramUser: response.external_name } );
 
-export default function InstagramGalleryEdit( props ) {
-	const { attributes, className, setAttributes } = props;
-	const { columns, photosPadding, photosToShow } = attributes;
+			// If Instagram sends back less images than the requested amount,
+			// this prevents showing placeholder for non existent images.
+			// Note: this will trigger an additional Instagram call.
+			if ( response.images.length < count ) {
+				setAttributes( { count: response.images.length } );
+			}
+		} );
+	}, [ accessToken, count ] );
+
+	const connectToInstagram = () => {
+		setIsConnecting( true );
+		apiFetch( { path: '/wpcom/v2/instagram/connect-url' } ).then( connectUrl => {
+			const popupMonitor = new PopupMonitor();
+
+			popupMonitor.open(
+				connectUrl,
+				'connect-to-instagram-popup',
+				'toolbar=0,location=0,menubar=0,' + popupMonitor.getScreenCenterSpecs( 700, 700 )
+			);
+
+			popupMonitor.on( 'message', ( { keyring_id } ) => {
+				setIsConnecting( false );
+				if ( keyring_id ) {
+					setAttributes( { accessToken: keyring_id.toString() } );
```

Comment:
In the widget we do a check at this point to [check if the current user has permissions to use the token_id](https://opengrok.a8c.com/source/xref/trunk/wp-content/mu-plugins/widgets/instagram/instagram.php#147) that is returned via ajax - do we need to do this here as well?

Reply:
If I read it correctly, the widget checks if the token is associated to the current user in Keyring.
I think it's a sensible thing to do, but as you say, once the token is saved as a block attribute, it becomes kinda public (well, in this case public to other authors of the site).
An alternative would be to not store it in the widget, but, for example, in a site option (which though would make the block more complex to write). 🤔

Do you have any opinions in merit?

Reply:
> Do you have any opinions in merit?

I had a brief discussion with Paul about this and one option we thought of was to check on the front end code that one of the site admins had permission to use the token - not sure how easy this is to do. 

We could also add a post content filter that did the same thing on post save, but I wonder if that is worth it - I am picking that the key thing is to stop someone displaying another person's feed on their front end without permission.

Reply:
Something I thought of here was storing this encrypted or signed with the Instagram client secret. The server could verify it then. That wouldn't stop someone copying the token around though, so you might want to also check that the token belongs to a member of the site.